### DISCUSSION TITLE
add support for trailing * to entityNamesFilter in pac modelbuilder

### DISCUSSION THREAD

**Discussion by danosted** [microsoft/powerplatform-build-tools#618] [Ideas]
When using pac modelbuilder with **entityNamesFilter** parameter it would be immensely useful to have support for wildcard character * for specific prefix models - e.g. "custom_*". This makes it possible to bulk add custom prefix models, instead of having to manually add all of a certain prefix.
It is the exact same idea as is implemented for **messagenamesfilter**.

**Comment by petrochuk** [comment]
It would be better to open the discussion here: https://github.com/microsoft/PowerPlatform-DataverseModelBuilder

Our tooling uses this library.

**Comment by JimDaly** [comment]
> Removed comment saying it worked. It doesn't.

**Reply by nicreuss** [reply]
I am also experiencing issues using the wildcard "*" to generate classes for tables with a common customization prefix. I am using the builderSettings.json approach with the pac cli.

If I modify the filter provided in the article as below (notice "contact" was replaced with "contac*", the only entity that is detected is the "account" entity.
```json
    "entityNamesFilter": [
        "account",
        "contac*",
        "sample_*"
    ]
```

Here is an excerpt from the cli output. Notice how 1 entity is read instead of 2.
```
Begin reading metadata from MetadataProviderService
        Begin Reading Metadata from Server
        Read 1 Entities - 00:00:00.653
        Read 0 Global OptionSets - 00:00:00.000
        Read 12 SDK Messages - 00:00:00.942
        Completed Reading Metadata from Server - 00:00:01.701
Completed reading metadata from MetadataProviderService - 00:00:01.704
Begin Writing Code Files
        Processing 1 Entities
        Wrote 1 Entities - 00:00:00.0487352
        Processing 12 Messages
        Wrote 2 Message(s). Skipped 10 Message(s) - 00:00:00.0108287
        Processing 0 Global OptionSets
        Wrote 0 Global OptionSets - 00:00:00.0000188
```

My thinking is that if the wildcard filter was working correctly, the output should be the same as the one given in your article.

Are you able to replicate this behaviour or offer any guidance?

**Reply by JimDaly** [reply]
Thanks for clarifying that this is an issue with direct command line and the builderSettings.json approach.
As @petrochuk suggested, it is better to raise this in https://github.com/microsoft/PowerPlatform-DataverseModelBuilder.

**Reply by nicreuss** [reply]
OK great thanks for the confirmation! Perhaps it would be beneficial to amend the article in the interim to prevent further readers from thinking this functionality exists for entities.

**Reply by JimDaly** [reply]
Better to spend time fixing it than documenting that it is broken.

**Reply by joe-oli** [reply]
Just to confirm it does NOT work, in fact I read the article that @JimDaly mentioned which suggested that wildcard functionality exits, but in practice that's not the case, so I came here looking for answers.

**Reply by JimDaly** [reply]
I just tested this again using the same `entityNamesFilter` I tested earlier:

```
  "entityNamesFilter": [
    "account",
    "contact",
    "sample_*"
  ],
```

This time it didn't work.

But the `messageNamesFilter` works fine.

I opened this issue [Wildcard characters on entityNamesFilter aren't working](https://github.com/microsoft/PowerPlatform-DataverseModelBuilder/issues/16) on the PowerPlatform-DataverseModelBuilder repo.

**Reply by JimDaly** [reply]
I was wrong. This never worked. I've updated [Generate early-bound classes for the SDK for .NET](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/org-service/generate-early-bound-classes) to remove the parts that said it worked.

**Comment by daryllabar** [comment]
@danosted this ability is supported in the Early Bound Generator V2 for the XrmToolBox:
![image](https://github.com/microsoft/powerplatform-build-tools/assets/9728423/1d9455ba-765c-498d-a38c-a3f988a19e0d)
This uses the PAC CLI under the covers, but provides additional features, like this one that you seek.

**Comment by JimDaly** [comment]
@danosted & @joe-oli 

As a workaround you can use with PowerShell & PAC CLI, this PowerShell function returns an array of table logical names that match an array of prefix values:

```
. $PSScriptRoot\Core.ps1

function Get-TableNamesFromPrefix {
   param (
      [Parameter(Mandatory)] 
      [String[]] 
      $prefixes
   )
   # Get table definitions
   $query = 'EntityDefinitions?$select=LogicalName'
   $tables = (Invoke-RestMethod `
         -Uri ($baseURI + $query) `
         -Method Get `
         -Headers $baseHeaders).value
   # The table names to return
   $tableNames = @()

   foreach ($table in $tables) {
      $logicalName = $table.LogicalName
      # some tables should be ignored
      if (!$logicalName.StartsWith('new_system_donotuseentity')) {
         foreach ($prefix in $prefixes) {
            if ($logicalName.StartsWith($prefix)) {
               $tableNames += $logicalName
            }
         }
      }
   }
   return $tableNames
}
```
For example:

```
Connect 'https://yourorg.crm.dynamics.com/' # change this
$names =  Get-TableNamesFromPrefix @('sample_','new_')
```

It is based on the pattern described in this article: [Use PowerShell and Visual Studio Code with the Dataverse Web API > Create reusable functions](https://learn.microsoft.com/power-apps/developer/data-platform/webapi/use-ps-and-vscode-web-api#create-reusable-functions)

@petrochuk Since [model-builder has already said they will not support this at this time](https://github.com/microsoft/PowerPlatform-DataverseModelBuilder/issues/16), perhaps PAC CLI can include this until they do?
